SU2_CFD/src/numerics_direct_mean.cpp

```diff
@@ -599,190 +359,151 @@ CUpwCUSP_Flow::CUpwCUSP_Flow(unsigned short val_nDim, unsigned short val_nVar, C
-    GetInviscidProjJac(MeanVelocity, &MeanEnergy, Normal, 0.5, val_Jacobian_i);
-    for (iVar = 0; iVar < nVar; iVar++)
-      for (jVar = 0; jVar < nVar; jVar++)
-        val_Jacobian_j[iVar][jVar] = val_Jacobian_i[iVar][jVar];
+  for (iDim = 0; iDim < nDim; iDim++) {
+    su2double MeanVel = (R*Velocity_j[iDim]+Velocity_i[iDim])/(R+1.0);
+    ProjVelocity += MeanVel*UnitNormal[iDim];
+    sq_vel += MeanVel*MeanVel;
   }
+  su2double MeanEnthalpy = (R*Enthalpy_j+Enthalpy_i)/(R+1.0);
+  su2double MeanSoundSpeed = sqrt(Gamma_Minus_One*fabs(MeanEnthalpy-0.5*sq_vel));
   
-  /*--- Computes differences conservative variables,
-   with a correction for the enthalpy ---*/
-  
-  for (iVar = 0; iVar < nVar; iVar++)
-    Diff_U[iVar] = U_i[iVar]-U_j[iVar];
-  Diff_U[nVar-1] = Density_i*Enthalpy_i-Density_j*Enthalpy_j;
-  
-  /*--- Computes differences projected fluxes,
-   with a correction for the enthalpy ---*/
-  
-  GetInviscidProjFlux(&Density_i, Velocity_i, &Pressure_i, &Enthalpy_i, UnitNormal, ProjFlux_i);
-  GetInviscidProjFlux(&Density_j, Velocity_j, &Pressure_j, &Enthalpy_j, UnitNormal, ProjFlux_j);
-  
-  for (iVar = 0; iVar < nVar; iVar++)
-    Diff_Flux[iVar] = ProjFlux_i[iVar]-ProjFlux_j[iVar];
+  su2double Mach = ProjVelocity / MeanSoundSpeed;
   
-  /*--- Compute dissipation parameters ---*/
+  su2double tmp1 = 0.5*(Gamma+1.0)/Gamma*ProjVelocity;
+  su2double tmp2 = sqrt(pow(tmp1-ProjVelocity/Gamma, 2.0) + pow(MeanSoundSpeed,2.0)/Gamma);
+  su2double LamdaNeg = tmp1 - tmp2, LamdaPos = tmp1 + tmp2;
   
-  Mach = ProjVelocity / MeanSoundSpeed;
-  
-  LamdaNeg = ProjVelocity - MeanSoundSpeed;
-  LamdaPos = ProjVelocity + MeanSoundSpeed;
-  
-  if ((0.0 <= Mach) && (Mach < 1.0)) Beta = + max(0.0, (ProjVelocity + LamdaNeg)/(ProjVelocity - LamdaNeg));
-  if ((-1.0 <= Mach) && (Mach < 0.0)) Beta = - max(0.0, (ProjVelocity + LamdaPos)/(ProjVelocity - LamdaPos));
   if (fabs(Mach) >= 1.0) Beta = Mach/fabs(Mach);
+  else if (Mach  >= 0.0) Beta = max(0.0, (ProjVelocity + LamdaNeg)/(ProjVelocity - LamdaNeg));
+  else                   Beta =-max(0.0, (ProjVelocity + LamdaPos)/(ProjVelocity - LamdaPos));
   
-  if (Beta == 0.0) Nu_c = fabs(ProjVelocity);
-  if ((Beta > 0.0) && ((0.0 < Mach) && (Mach < 1.0))) Nu_c = - (1.0+Beta)*LamdaNeg;
-  if ((Beta < 0.0) && ((-1.0 < Mach) && (Mach < 0.0))) Nu_c = (1.0-Beta)*LamdaPos;
-  if (fabs(Mach) >= 1) Nu_c = 0.0;
+  if (fabs(Mach) >= 1.0) Nu_c = 0.0;
+  else {
+    if      (Beta > 0.0) Nu_c =-(1.0+Beta)*LamdaNeg;
+    else if (Beta < 0.0) Nu_c = (1.0-Beta)*LamdaPos;
+    /*--- Limit the minimum scalar dissipation ---*/
+    else Nu_c = max(fabs(ProjVelocity), config->GetEntropyFix_Coeff()*MeanSoundSpeed);
+  }  
   
   /*--- Compute the residual ---*/
   
   for (iVar = 0; iVar < nVar; iVar++)
-    val_residual[iVar] += (0.5*Nu_c*Diff_U[iVar] + 0.5*Beta*Diff_Flux[iVar])*Area;
+    val_residual[iVar] = 0.5*((1.0+Beta)*ProjFlux_i[iVar] + (1.0-Beta)*ProjFlux_j[iVar] + Nu_c*Diff_U[iVar])*Area;
 
```

Comment:
As discussed in #705, according to the literature the CUSP flux should be expressed as the average of the left and right fluxes plus the artificial diffusion part, and not as a flux of averages + diffusion. This way the scheme becomes 100% upwind above Mach 1, and as a bonus the implementation also becomes more efficient.

Whatever I could do to reduce duplication and improve convergence is done, this PR is ready for reviews.